Restore needed thcovmat functions

andreab1997 opened the pull request:
It seems that #1899 removed some functions that are needed for some of the reports we usually do (cc @giacomomagni). For example for `shift_diag_cov_comparison`. I am trying to restore and eventually fix this here. @RoyStegeman @scarlehoff

scarlehoff commented on validphys2/src/validphys/theorycovariance/construction.py:
```diff
@@ -171,6 +171,37 @@ def combine_by_type(each_dataset_results_central_bytheory):
+    to the starting point of each process."""
+    process_info = combine_by_type
+    running_index = 0
+    start_proc = defaultdict(list)
```
why is this a list?

scarlehoff commented on validphys2/src/validphys/theorycovariance/construction.py:
```diff
@@ -171,6 +171,37 @@ def combine_by_type(each_dataset_results_central_bytheory):
+    process_info = combine_by_type
+    running_index = 0
+    start_proc = defaultdict(list)
+    for name in process_info.theory:
+        size = len(process_info.theory[name][0])
```
```suggestion
    for name, thdata in process_info.theory.items():
        # Get the total number of points of the process
        size = len(thdata[0])
```

scarlehoff commented on validphys2/src/validphys/theorycovariance/construction.py:
```diff
@@ -171,6 +171,37 @@ def combine_by_type(each_dataset_results_central_bytheory):
+    return start_proc
+
+
+def covmap(combine_by_type, dataset_names):
```
Instead of covmat please create a function `dataset_startingpoints` like the one for process.

scarlehoff commented on validphys2/src/validphys/theorycovariance/construction.py:
```diff
@@ -171,6 +171,37 @@ def combine_by_type(each_dataset_results_central_bytheory):
+    for dataset in dataset_names:
+        size = process_info.sizes[dataset]
+        start_exp[dataset] = running_index
+        running_index += size
```
I think this risks the same bug as before (I haven't tested it since I'm in the phone but I think it might be).

However, at this point (and also covmat by dataset) you always have the theory covmat so there's no need to use combine_by_type and simply hope for the best.

Rather you should look at the index of the theory covmat (as a dataframe) and select the index you want.

I would also argue that you don't need a mapping if you have a dataframe because you can simply ask for the first index of process X or dataset Y.

So the actual solution is, for the functions that use the theory (or experimental) covmat as an array, to use them as a dataframe.

If that's beyond the realm of possibility right now for whatever reason, this function should take also a covmat df as input (and output that same covmat together with the mapping) to avoid any ambigüities

scarlehoff reviewed (dismissed):
I guess the important question is, do we need this for the papers and do we need this quick?

I might be able to work tomorrow on changing the functions that use arrays to use dataframes instead (if you point me to them).

As I said in the comments, I'm not convinced `covmap` is correct unless the order in the runcard is the "right one".

edit: note that the need for all this trickery was probably due to the mix of C++ and python. Now that in vp we can have dataframes all way through you should not need any mappings because the dataframe contains the mapping.

giacomomagni commented:
> I guess the important question is, do we need this for the papers and do we need this quick?
>
Unfortunately I think we need this plot (ie the red and black shift plot) for the papers. 😭 
But at this points, fits are delayed after Christmas, so we have a bit of time. 

Thanks you all for the help.

scarlehoff commented:
Only `shift_diag_cov_comparison` is needed? (and currently broken)

giacomomagni commented:
> Only `shift_diag_cov_comparison` is needed? (and currently broken)

I think so, running the previous vp-runcard I'm getting: 

```
A parameter is required: theory_covmat_custom_dataspecs.
This is needed to process:
 - report
through:
 - template_text
through:
 - shift_diag_cov_comparison
through:
 - thx_covmat
 ```

RoyStegeman commented:
Can you share your runcard?

giacomomagni commented:
[template_shifts_n3lo.md](https://github.com/NNPDF/nnpdf/files/13742694/template_shifts_n3lo.md)

<details>
<summary> here you go: </summary>

```
meta:
   author: Giacomo Magni
   keywords: [theory uncertainties, 7-point, N3LO]
   title: validation shifts NNLO vs N3LO for DIS and DY 

datasets_inputs_nnlo: &datasets_list_nnlo
  - {dataset: NMCPD_dw_ite}
  - {dataset: NMC}
  - {dataset: SLACP_dwsh}
  - {dataset: SLACD_dw_ite}
  - {dataset: BCDMSP_dwsh}
  - {dataset: BCDMSD_dw_ite}
  - {dataset: CHORUSNUPb_dw_ite}
  - {dataset: CHORUSNBPb_dw_ite}
  - {dataset: NTVNUDMNFe_dw_ite, cfac: [MAS]}
  - {dataset: NTVNBDMNFe_dw_ite, cfac: [MAS]}
  - {dataset: HERACOMBNCEM}
  - {dataset: HERACOMBNCEP460}
  - {dataset: HERACOMBNCEP575}
  - {dataset: HERACOMBNCEP820}
  - {dataset: HERACOMBNCEP920}
  - {dataset: HERACOMBCCEM}
  - {dataset: HERACOMBCCEP}
  - {dataset: HERACOMB_SIGMARED_C}
  - {dataset: HERACOMB_SIGMARED_B}
  # - {dataset: DYE886R_dw_ite}
  # - {dataset: DYE886P, cfac: []}
  #- {dataset: DYE605_dw_ite, cfac: []}
  #- {dataset: DYE906R_dw_ite, cfac: [ACC]}
  - {dataset: CDFZRAP_NEW}
  - {dataset: D0ZRAP_40, cfac: []}
  - {dataset: D0WMASY}
  - {dataset: ATLASWRAP36PB}
  - {dataset: ATLASZRAP36PB}
  - {dataset: ATLASZHIGHMASS49FB}
  - {dataset: ATLASLOMASSDY11EXT}
  - {dataset: ATLASWRAP11CC}
  - {dataset: ATLASZRAP11CC}
  - {dataset: ATLASWZRAP11CF}
  - {dataset: ATLASDY2D8TEV, cfac: []}
  - {dataset: ATLAS_DY_2D_8TEV_LOWMASS, cfac: []}
  - {dataset: ATLAS_Z_TOT_13TEV, cfac: [NRM]}
  - {dataset: ATLAS_W_TOT_13TEV, cfac: [NRM]}
  # - {dataset: ATLAS_WP_JET_8TEV_PT, cfac: []}
  # - {dataset: ATLAS_WM_JET_8TEV_PT, cfac: []}
  # #- {dataset: ATLAS_WCHARM_WP_DIFF_7TEV, cfac: []}
  # #- {dataset: ATLAS_WCHARM_WM_DIFF_7TEV, cfac: []}
  - {dataset: ATLASZPT8TEVMDIST, cfac: []}
  - {dataset: ATLASZPT8TEVYDIST, cfac: []}
  # - {dataset: ATLASTTBARTOT7TEV, cfac: []}
  # - {dataset: ATLASTTBARTOT8TEV, cfac: []}
  # - {dataset: ATLAS_TTBARTOT_13TEV_FULLLUMI, cfac: []}
  # - {dataset: ATLAS_TTB_DIFF_8TEV_LJ_TRAPNORM, cfac: []}
  # - {dataset: ATLAS_TTB_DIFF_8TEV_LJ_TTRAPNORM, cfac: []}
  # - {dataset: ATLAS_TOPDIFF_DILEPT_8TEV_TTRAPNORM, cfac: []}
  # - {dataset: ATLAS_1JET_8TEV_R06_DEC, cfac: []}
  # - {dataset: ATLAS_2JET_7TEV_R06, cfac: []}
  # - {dataset: ATLASPHT15_SF, cfac: [EWK]}
  # - {dataset: ATLAS_SINGLETOP_TCH_R_7TEV, cfac: []}
  # - {dataset: ATLAS_SINGLETOP_TCH_R_13TEV, cfac: []}
  # - {dataset: ATLAS_SINGLETOP_TCH_DIFF_7TEV_T_RAP_NORM, cfac: []}
  # - {dataset: ATLAS_SINGLETOP_TCH_DIFF_7TEV_TBAR_RAP_NORM, cfac: []}
  # - {dataset: ATLAS_SINGLETOP_TCH_DIFF_8TEV_T_RAP_NORM, cfac: []}
  # - {dataset: ATLAS_SINGLETOP_TCH_DIFF_8TEV_TBAR_RAP_NORM, cfac: []}
  - {dataset: CMSWEASY840PB}
  - {dataset: CMSWMASY47FB}
  - {dataset: CMSDY2D11, cfac: []}
  - {dataset: CMSWMU8TEV}
  - {dataset: CMSZDIFF12, cfac: [NRM]}
  # - {dataset: CMS_2JET_7TEV, cfac: []}
  # - {dataset: CMS_1JET_8TEV, cfac: []}
  # - {dataset: CMSTTBARTOT7TEV, cfac: []}
  # - {dataset: CMSTTBARTOT8TEV, cfac: []}
  # - {dataset: CMSTTBARTOT13TEV, cfac: []}
  # - {dataset: CMSTOPDIFF8TEVTTRAPNORM, cfac: []}
  # - {dataset: CMSTTBARTOT5TEV, cfac: []}
  # - {dataset: CMS_TTBAR_2D_DIFF_MTT_TRAP_NORM, cfac: []}
  # - {dataset: CMS_TTB_DIFF_13TEV_2016_2L_TRAP, cfac: []}
  # - {dataset: CMS_TTB_DIFF_13TEV_2016_LJ_TRAP, cfac: []}
  # - {dataset: CMS_SINGLETOP_TCH_TOT_7TEV, cfac: []}
  # - {dataset: CMS_SINGLETOP_TCH_R_8TEV, cfac: []}
  # - {dataset: CMS_SINGLETOP_TCH_R_13TEV, cfac: []}
  #- {dataset: CMSWCHARMTOT}
  #- {dataset: CMSWCHARMRAT}
  #- {dataset: CMS_WCHARM_DIFF_UNNORM_13TEV}
  - {dataset: LHCBZ940PB}
  - {dataset: LHCBZEE2FB_40}
  - {dataset: LHCBWMU7TEV, cfac: [NRM]}
  - {dataset: LHCBZMU7TEV, cfac: [NRM]}
  - {dataset: LHCBWMU8TEV, cfac: [NRM]}
  - {dataset: LHCBZMU8TEV, cfac: [NRM]}
  - {dataset: LHCB_Z_13TEV_DIMUON}
  - {dataset: LHCB_Z_13TEV_DIELECTRON}

datasets_inputs_n3lo: &datasets_list_n3lo
  - {dataset: NMCPD_dw_ite}
  - {dataset: NMC}
  - {dataset: SLACP_dwsh}
  - {dataset: SLACD_dw_ite}
  - {dataset: BCDMSP_dwsh}
  - {dataset: BCDMSD_dw_ite}
  - {dataset: CHORUSNUPb_dw_ite}
  - {dataset: CHORUSNBPb_dw_ite}
  - {dataset: NTVNUDMNFe_dw_ite, cfac: [MAS]}
  - {dataset: NTVNBDMNFe_dw_ite, cfac: [MAS]}
  - {dataset: HERACOMBNCEM}
  - {dataset: HERACOMBNCEP460}
  - {dataset: HERACOMBNCEP575}
  - {dataset: HERACOMBNCEP820}
  - {dataset: HERACOMBNCEP920}
  - {dataset: HERACOMBCCEM}
  - {dataset: HERACOMBCCEP}
  - {dataset: HERACOMB_SIGMARED_C}
  - {dataset: HERACOMB_SIGMARED_B}
  # - {dataset: DYE886R_dw_ite}
  # - {dataset: DYE886P, cfac: []}
  #- {dataset: DYE605_dw_ite, cfac: []}
  #- {dataset: DYE906R_dw_ite, cfac: [ACC]}
  - {dataset: CDFZRAP_NEW, cfac: []}
  - {dataset: D0ZRAP_40, cfac: []}
  - {dataset: D0WMASY, cfac: []}
  - {dataset: ATLASWRAP36PB, cfac: []}
  - {dataset: ATLASZRAP36PB, cfac: []}
  - {dataset: ATLASZHIGHMASS49FB, cfac: []}
  - {dataset: ATLASLOMASSDY11EXT, cfac: []}
  - {dataset: ATLASWRAP11CC, cfac: []}
  - {dataset: ATLASZRAP11CC, cfac: []}
  - {dataset: ATLASWZRAP11CF, cfac: []}
  - {dataset: ATLASDY2D8TEV, cfac: []}
  - {dataset: ATLAS_DY_2D_8TEV_LOWMASS, cfac: []}
  - {dataset: ATLAS_Z_TOT_13TEV, cfac: [NRM,]}
  - {dataset: ATLAS_W_TOT_13TEV, cfac: [NRM,]}
  # - {dataset: ATLAS_WP_JET_8TEV_PT, cfac: []}
  # - {dataset: ATLAS_WM_JET_8TEV_PT, cfac: []}
  # #- {dataset: ATLAS_WCHARM_WP_DIFF_7TEV, cfac: []}
  # #- {dataset: ATLAS_WCHARM_WM_DIFF_7TEV, cfac: []}
  - {dataset: ATLASZPT8TEVMDIST, cfac: []}
  - {dataset: ATLASZPT8TEVYDIST, cfac: []}
  # - {dataset: ATLASTTBARTOT7TEV, cfac: []}
  # - {dataset: ATLASTTBARTOT8TEV, cfac: []}
  # - {dataset: ATLAS_TTBARTOT_13TEV_FULLLUMI, cfac: []}
  # - {dataset: ATLAS_TTB_DIFF_8TEV_LJ_TRAPNORM, cfac: []}
  # - {dataset: ATLAS_TTB_DIFF_8TEV_LJ_TTRAPNORM, cfac: []}
  # - {dataset: ATLAS_TOPDIFF_DILEPT_8TEV_TTRAPNORM, cfac: []}
  # - {dataset: ATLAS_1JET_8TEV_R06_DEC, cfac: []}
  # - {dataset: ATLAS_2JET_7TEV_R06, cfac: []}
  # - {dataset: ATLASPHT15_SF, cfac: [EWK]}
  # - {dataset: ATLAS_SINGLETOP_TCH_R_7TEV, cfac: []}
  # - {dataset: ATLAS_SINGLETOP_TCH_R_13TEV, cfac: []}
  # - {dataset: ATLAS_SINGLETOP_TCH_DIFF_7TEV_T_RAP_NORM, cfac: []}
  # - {dataset: ATLAS_SINGLETOP_TCH_DIFF_7TEV_TBAR_RAP_NORM, cfac: []}
  # - {dataset: ATLAS_SINGLETOP_TCH_DIFF_8TEV_T_RAP_NORM, cfac: []}
  # - {dataset: ATLAS_SINGLETOP_TCH_DIFF_8TEV_TBAR_RAP_NORM, cfac: []}
  - {dataset: CMSWEASY840PB, cfac: []}
  - {dataset: CMSWMASY47FB, cfac: []}
  - {dataset: CMSDY2D11, cfac: []}
  - {dataset: CMSWMU8TEV, cfac: []}
  - {dataset: CMSZDIFF12, cfac: [NRM]}
  # - {dataset: CMS_2JET_7TEV, cfac: []}
  # - {dataset: CMS_1JET_8TEV, cfac: []}
  # - {dataset: CMSTTBARTOT7TEV, cfac: []}
  # - {dataset: CMSTTBARTOT8TEV, cfac: []}
  # - {dataset: CMSTTBARTOT13TEV, cfac: []}
  # - {dataset: CMSTOPDIFF8TEVTTRAPNORM, cfac: []}
  # - {dataset: CMSTTBARTOT5TEV, cfac: []}
  # - {dataset: CMS_TTBAR_2D_DIFF_MTT_TRAP_NORM, cfac: []}
  # - {dataset: CMS_TTB_DIFF_13TEV_2016_2L_TRAP, cfac: []}
  # - {dataset: CMS_TTB_DIFF_13TEV_2016_LJ_TRAP, cfac: []}
  # - {dataset: CMS_SINGLETOP_TCH_TOT_7TEV, cfac: []}
  # - {dataset: CMS_SINGLETOP_TCH_R_8TEV, cfac: []}
  # - {dataset: CMS_SINGLETOP_TCH_R_13TEV, cfac: []}
  #- {dataset: CMSWCHARMTOT}
  #- {dataset: CMSWCHARMRAT}
  #- {dataset: CMS_WCHARM_DIFF_UNNORM_13TEV}
  - {dataset: LHCBZ940PB, cfac: []}
  - {dataset: LHCBZEE2FB_40, cfac: []}
  - {dataset: LHCBWMU7TEV, cfac: [NRM,]}
  - {dataset: LHCBZMU7TEV, cfac: [NRM]}
  - {dataset: LHCBWMU8TEV, cfac: [NRM,]}
  - {dataset: LHCBZMU8TEV, cfac: [NRM,]}
  - {dataset: LHCB_Z_13TEV_DIMUON, cfac: []}
  - {dataset: LHCB_Z_13TEV_DIELECTRON, cfac: []}


orthonormalisation: qr # Choice of orthonormalisation scheme
                             # for finding th covmat basis. Default is qr.
theoryid: 708
      
pdf: "NNPDF40_nnlo_as_01180"
use_cuts: internal
q2min: 3.49                       # Q2 minimum
w2min: 12.5                       # W2 minimum


shiftconfig:     # For calculating NNLO-NLO shift

  use_cuts: internal
  q2min: 3.49                      # Q2 minimum
  w2min: 12.5                       # W2 minimum


  theoryid: 708
      
  dataspecs:
    - theoryid: 708
      pdf: "NNPDF40_nnlo_as_01180"
      speclabel: "NNLO"
      dataset_inputs: *datasets_list_nnlo
                          
    - theoryid: 722
      pdf: "NNPDF40_nnlo_as_01180"
      speclabel: "N3LO"
      dataset_inputs: *datasets_list_n3lo

theoryconfig:

  theoryid: 708

  theoryids:
    from_: scale_variation_theories

  point_prescription: '7 point'
  use_cuts: internal
  q2min: 3.49                      # Q2 minimum
  w2min: 12.5                       # W2 minimum
  pdf: "NNPDF40_nnlo_as_01180"     

  dataspecs:
    - theoryid: 708
      speclabel: $(\xi_F,\xi_R)=(1,1)$
      dataset_inputs: *datasets_list_nnlo
    - theoryid: 711
      speclabel: $(\xi_F,\xi_R)=(2,1)$
      dataset_inputs: *datasets_list_nnlo
    - theoryid: 705
      speclabel: $(\xi_F,\xi_R)=(0.5,1)$
      dataset_inputs: *datasets_list_nnlo
    - theoryid: 709
      speclabel: $(\xi_F,\xi_R)=(1,2)$
      dataset_inputs: *datasets_list_nnlo
    - theoryid: 707
      speclabel: $(\xi_F,\xi_R)=(1,0.5)$
      dataset_inputs: *datasets_list_nnlo
    - theoryid: 712
      speclabel: $(\xi_F,\xi_R)=(2,2)$
      dataset_inputs: *datasets_list_nnlo
    - theoryid: 704
      speclabel: $(\xi_F,\xi_R)=(0.5,0.5)$
      dataset_inputs: *datasets_list_nnlo

template: template_shifts_n3lo.md

dataset_report:
   meta: Null
   template_text: |
      ## Testing 7pt NNLO global covariance matrix against N3LO-NNLO shift
actions_:
  - report(main=true, mathjax=True)
```

RoyStegeman commented:
Do we know why is the theorycovmat shift calculated using the diagonal elements of the theory covmat instead of e.g. the values on the diagonal of the covmat with all variations considered? I know both neglect part of the picture and I'm not sure which is the better choice for interpretation/presentation, but do you know the motivation for this chioce?

giacomomagni commented:
> Do we know why is the theorycovmat shift calculated using the diagonal elements of the theory covmat instead of e.g. the values on the diagonal of the covmat with all variations considered? 

Personally I don't, but this looks more something that happens when you start with a basic setup and the you add features and you forgot to update everything.

> I know both neglect part of the picture and I'm not sure which is the better choice for interpretation/presentation, but do you know the motivation for this chioce?

I'd say we can fix this behaviour and always compute the full covmat. If only the diagonal is needed you extract them from the full matrix.  Maybe @andreab1997 or someone else have better ideas.

RoyStegeman commented:
> I'd say we can fix this behaviour and always compute the full covmat. If only the diagonal is needed you extract them from the full matrix. Maybe @andreab1997 or someone else have better ideas.

I was indeed thinking along those lines

andreab1997 commented:
> Only `shift_diag_cov_comparison` is needed? (and currently broken)

No we need all the functions in `tests.py` (at least I need all of them).

andreab1997 commented:
Just to keep track of this: Now the function `shift_diag_cov_comparison_test` works and does what expected. However while in some cases the results are very similar to the old ones (as expected), for example in the `DIS NC` case, in other cases the results are pretty weird, for instance in the `DY CC` case. Given that now the dataset order is enforced trought the DataFrames, I fear there might be a problem with the order of the datapoints in the datasets and/or with the labels of the theory_covmat. @RoyStegeman and @scarlehoff when you can, please have a look.

I leave here a couple of runcard that you can use (given that I simplified their structure a bit):
DIS NC: [231116-valplot-DISNC.txt](https://github.com/NNPDF/nnpdf/files/13861042/231116-valplot-DISNC.txt)
DY CC: [231116-valplot-DYCC.txt](https://github.com/NNPDF/nnpdf/files/13861045/231116-valplot-DYCC.txt)
template: [template_test.md](https://github.com/NNPDF/nnpdf/files/13861017/template_test.md)

andreab1997 commented:
Ok maybe the problem was just the one corrected in [52f3262](https://github.com/NNPDF/nnpdf/pull/1901/commits/52f32621b41f2769badafb36e9c322cb5d670cbf) but if you can please have a look anyway

RoyStegeman commented:
> Given that now the dataset order is enforced trought the DataFrames, I fear there might be a problem with the order of the datapoints in the datasets and/or with the labels of the theory_covmat

Do you mean the problem is with the labels of the stored covmat from vp-setupfit or only in these plotting function? I'm not so sure what you'd like us to check, could you share the plot/table you're worried about?

andreab1997 commented:
> > Given that now the dataset order is enforced trought the DataFrames, I fear there might be a problem with the order of the datapoints in the datasets and/or with the labels of the theory_covmat
> 
> Do you mean the problem is with the labels of the stored covmat from vp-setupfit or only in these plotting function? I'm not so sure what you'd like us to check, could you share the plot/table you're worried about?

No, after [52f3262](https://github.com/NNPDF/nnpdf/pull/1901/commits/52f32621b41f2769badafb36e9c322cb5d670cbf) maybe there is no problem. For example this is the new report for [DYCC](https://vp.nnpdf.science/W_nQh15CQv-Az4aYFutzAw==) to be compared with the old one [DYCC](https://vp.nnpdf.science/pguRHnvoTnW1DMHQ8ZRHUQ==/). Of course they are not the same and they should not given that the old one was affected by all the ordering problems, but now they look reasonably similar. Anyway, I will ask you for a review but if you want to have a look at the code even now, please do it even if for the moment it is just a draft.

giacomomagni commented on validphys2/src/validphys/theorycovariance/tests.py:
```diff
@@ -766,3 +800,52 @@ def shift_diag_cov_comparison(allthx_vector, shx_vector, thx_covmat, thx_vector)
+    fig, ax = plotutils.subplots(figsize=(20, 10))
+    ax.plot(np.array(sqrtdiags_concat) * 100, ".-", label=f"MHOU ({point_prescription})", color="red")
+    ax.plot(-np.array(sqrtdiags_concat) * 100, ".-", color="red")
+    ax.plot(-np.array(fnorm_concat) * 100, ".-", label="NNLO-NLO Shift", color="black")
```
Just to remember, to change the label for different PTOs

giacomomagni commented on validphys2/src/validphys/theorycovariance/tests.py:
```diff
@@ -766,3 +800,52 @@ def shift_diag_cov_comparison(allthx_vector, shx_vector, thx_covmat, thx_vector)
+    startlocs_lines = [x - 0.5 for x in startlocs]
+    ax.vlines(startlocs_lines, -70, 70, linestyles="dashed")
+    ax.margins(x=0, y=0)
+    ax.set_ylabel(r"% wrt central theory $T_i^{(0)}$", fontsize=20)
```
```suggestion
    ax.set_ylabel(r"PTO difference vs diagonal MHOU normalised to central value", fontsize=20)
```

andreab1997 replied:
[6491368](https://github.com/NNPDF/nnpdf/pull/1901/commits/6491368251e05cb7bff99b4b1343be3bb0290839)

giacomomagni reviewed (commented):
LGTM and it's working!

andreab1997 commented:
@scarlehoff I resolved all your conversations because they refer to functions that I now removed ([5db9ae0](https://github.com/NNPDF/nnpdf/pull/1901/commits/5db9ae053853fc5fe9dc92876d06f1abbcbc2654)), together with all the others that are not used anymore in the theorycovariance module. Please check if in [5db9ae0](https://github.com/NNPDF/nnpdf/pull/1901/commits/5db9ae053853fc5fe9dc92876d06f1abbcbc2654) I removed something that you use or that I should not have removed.

scarlehoff commented on validphys2/src/validphys/theorycovariance/tests.py:
```diff
@@ -399,41 +268,36 @@ def vectors_9pt(splitdiffs):
-    for index in indexlist:
-        name = index[0]
-        proc = process_lookup(name)
+    for index in doubleindex:
+        name = index[1]
+        proc = index[0]
```
```suggestion
    for proc, name in doubleindex:
```

scarlehoff commented on validphys2/src/validphys/theorycovariance/tests.py:
```diff
@@ -491,23 +355,26 @@ def evals_nonzero_basis(
-    return w, v, cond_num
-
+    return w, v, cond_num 
+    
```
Remember to run black

scarlehoff commented on validphys2/src/validphys/theorycovariance/tests.py:
```diff
@@ -665,42 +516,27 @@ def eigenvector_plot(evals_nonzero_basis, shx_vector):
-        [processnames, dsnames, ids], names=("process", "dataset", "id")
-    )
-    f = pd.DataFrame(f.values, index=tripleindex)
+    f = -pd.DataFrame(fnorm_concat, index=tripleindex_thcovmat_complete)
```
Please do `-1.0*`

scarlehoff reviewed (commented):
I'm completely happy with anything that implies removing code from this module :P

andreab1997 commented:
This is the complete test report that we need for the papers as obtained with the current status of this branch: https://vp.nnpdf.science/MLy8yUWdRyOqJpEYERiEmg== . Results are very similar to old ones but of course not completely identical given that we solved the ordering issue. I am now about to check and improve everything again in `tests.py` so comments and suggestions are welcomed (note however that at the moment this is really a draft).

scarlehoff commented:
> comments and suggestions are welcomed

Regression tests. Once you have a result which you know _for sure_ that it is correct, add a regression test. You can use the theories from here https://docs.nnpdf.science/vp/examples.html?highlight=resources#recommended-resources 
That way future changes that change the order in magical ways will make the tests fail.

andreab1997 commented:
Given that from my side everything is ready here (excluding regression tests), I would say that this is ready to be reviewed. 

@scarlehoff about regression tests I was thinking about comparing reports (given that the tests pourpose is exactly producing a validation report). How do you suggest to do?

RoyStegeman commented:
> @scarlehoff about regression tests I was thinking about comparing reports (given that the tests pourpose is exactly producing a validation report). How do you suggest to do?

Instead of comparing reports you could compare plots (or tables with info used to construct the plots)

scarlehoff commented:
Yes. I'd suggest comparing dataframes (there's a few examples in the tests). You can compare for instance theory covmat with a few edge cases that you know would fail for a problematic ordering.

RoyStegeman commented on validphys2/src/validphys/theorycovariance/construction.py:
```diff
@@ -31,37 +31,6 @@
-
-@table
-@check_correct_theory_combination
-def theory_covmat_singleprocess(theory_covmat_singleprocess_no_table, fivetheories):
```
This function is still used in other places in this module, so either this function should not be removed, or all those function relying on this should (recursively) be removed as well

RoyStegeman commented on validphys2/src/validphys/theorycovariance/construction.py:
```diff
@@ -98,44 +67,6 @@ def theory_covmat_dataset(
-
-
-@check_correct_theory_combination
-def total_covmat_procs(procs_results_theory, fivetheories):
```
Still used

RoyStegeman reviewed (changes requested):
Please make sure to not leave any dead code after removing functions. I only pointed out two of them, but I didn't check all functions so there may be more.

andreab1997 commented on validphys2/src/validphys/theorycovariance/tests.py:
```diff
@@ -716,45 +577,81 @@ def deltamiss_plot(theory_shift_test, allthx_vector, evals_nonzero_basis, shx_ve
+    """Plot of the comparison of a shift between two pertubative order and the diagonal entries of the theory covmat, both normalized to the first of the two perturbative orders."""
+    # Concatenation of the arrays
+    fnorm_concat = [j for i in fnorm_shifts_byprocess for j in i]
+    sqrtdiags_concat = [j for i in sqrtdiags_thcovmat_byprocess for j in i]
```
[3a428bf](https://github.com/NNPDF/nnpdf/pull/1901/commits/3a428bf55e4142b1eece4cd86f0d0a947b8d54eb)

RoyStegeman commented on validphys2/src/validphys/theorycovariance/construction.py:
```diff
@@ -703,179 +579,14 @@ def theory_normcovmat_custom(theory_covmat_custom, procs_data_values):
     plus theory covariance matrices, correlations by prescription."""
-    corrmat = experimentplustheory_corrmat_singleprocess(procs_covmat, theory_covmat_custom)
+    total_df = procs_covmat + theory_covmat_custom
+    total_cov = (procs_covmat + theory_covmat_custom).values
```
```suggestion
```

andreab1997 replied:
[f6263c2](https://github.com/NNPDF/nnpdf/pull/1901/commits/f6263c2df5ba571ce53b4c621e240e17464d65ee)

RoyStegeman reviewed (approved):
Okay I didn't look at it super close but given the results I'll trust that it works. We don't want to wait a long time with merging this either.

andreab1997 commented:
> Okay I didn't look at it super close but given the results I'll trust that it works. We don't want to wait a long time with merging this either.

Ok then I would say that if @giacomomagni can try again now his usual plots just to be sure, then we can merge. Note that I am about to do just another commit for cosmetic changes so wait for that.

andreab1997 commented:
> I'm happy with this PR, but unfortunately some changes are still required (at least for the N3LO):
> 
> In particular:
> 
> ```
> - Different marker style for each category
> - Avoid yellow as color
> ```
> 
> shall I give a try to implement them here?

I would say, let's merge this now in such a way we can tag the code and then let's open another PR for these changes you are asking for. Do you agree?
